### PR TITLE
Implement new query API

### PR DESCRIPTION
**Pull request by berfarah**
Depends on https://github.com/coverhound/active-redux/pull/32 - rebasing as they're approved

Closes https://github.com/coverhound/active-redux/issues/33 and draws heavy inspiration from [Ember's finding methods](https://guides.emberjs.com/v2.14.0/models/finding-records/)

In this PR
=======

Our new API is much clearer about what it's doing:

Methods with "peek" in the name will always check local store

    Model.peek()
    Model.peekAll()

These methods will always check remotely. 

    Model.find()
    Model.findAll()
    Model.query()

These methods will create query selectors. They serve up a local version if possible, while also making a request to the server.

    Model.select()
    Model.selectAll()
    Model.selectQuery()

The separation of query logic between store and queries has been cleaned up. Due to the clear lines in what API methods do now, the code is much clearer, and the separation is no longer necessary.

`Model#query`'s hashing for indices is based on what the endpoint call would look like. For example:

    Person.query({ page: 2 })
    // becomes `people?page=2`

A new dependency called `qs` handles the stringification.

Also in this PR
==============

As I was restructuring / improving tests, I came across various bugs - and bugs waiting to happen - which I fixed along the way.

###  Change relationships to be static by default

Having used the API, it seems fairly painful to try to navigate model
relationships with promises. They now default to returning what's in the
store (even if it's nothing).

Added methods to fetch relationships via `fetch[thing]` methods.

Example:

```js
import * as AR from 'active-redux';

const Company = AR.define('companies', class Company {});
const User = AR.define('users', class User {
  static attributes = {
    employer: AR.Attr.hasOne('companies'),
  };
});

async function() {
  const user = await User.find("5");
  user.employer // => Company|null
  user.fetchEmployer() // => Promise<Company|null>
}
```

### Implement selectors

Added a function to make this library fairly easy to use with `react-redux`:

```jsx
import { Comment } from '@models';

const View = ({ comment }) => <p>{comment.body}</p>;

// needed in order to have a unique selector per instance of the component
const mapStateToProps = () => {
  const commentFind = Comment.select;
  return (state, props) => ({
    // Arguments passed into querySelector are passed to `Comment#find`
    // and turned into a selector
    comments: commentFind(props.commentId)(state)
  });
};

const mapDispatchToProps = () => ({});

export default connect(mapStateToProps, mapDispatchToProps)(View);
```

### Added a working example

If you want to try this out, right now it's working off of symlinks. Going to change it to use `yarn` soonish.

To get it running:

```sh
cd example
yarn install
ln -s ../dist node_modules/active-redux

# in one window
yarn server

# in another
yarn start
```

### CODE REVIEW

File: src/indexing/index.js [hunk 1]
```diff
@@ -1 +1,12 @@
+import qs from 'qs';
+
 export extension from './reducer';
+export * from './reducer';
+
+export const queryEndpoint = (endpoint, query) => {
```
Comment (berfarah) [inline src/indexing/index.js]: This might belong in `src/api/utils`

File: src/indexing/reducer.js [hunk 2]
```diff
@@ -1,6 +1,22 @@
 import imm from 'object-path-immutable';
 import { createAction, resourcesArray } from '../api/utils';
 
+/**
+ * Helpers
+ */
+class Index extends Array {
```
Comment (berfarah) [inline src/indexing/reducer.js]: So... subclassing primitives is something you can do in node - but not Babel. So this only half-works. The constructor works, but any subsequent manipulations are on Array, not Index. 

I'm debating on whether this should just live as logic in `indexArray` to make it less confusing.

File: src/indexing/reducer.js [hunk 3]
```diff
@@ -22,37 +38,33 @@ export const apiIndexAsync = ({ hash, promise }) => (dispatch, getState) => {
...
+  return promise.then((data) => {
+    if (indexExists()) dispatch(apiIndexDone({ hash, data }));
+    return data;
+  }).catch((err) => {
+    dispatch(apiIndexDone({ hash, data: [] }));
+    return Promise.reject(err);
   });
 };
-export const apiIndexSync = ({ hash, resources }) => (dispatch) => {
+export const apiIndexSync = ({ hash, data }) => (dispatch) => {
```
Comment (berfarah) [inline src/indexing/reducer.js]: The distinction between `resources` and `data` was a bug waiting to happen

File: src/model.js [hunk 4]
```diff
@@ -28,12 +27,13 @@ const defineRelationship = (object, field, attribute) => {
...
   if (!model) throw new Error(`Unregistered model: ${resource}`);
-  const fetch = (isArray ? Queries.hasMany : Queries.hasOne)(model);
+  const find = ({ id } = {}) => Store.find(id, { model });
+  const fetch = (isArray ? (data = []) => Promise.all(data.map(find)) : find);
 
   Object.defineProperty(object, field, {
     get() {
       const data = this.data.relationships[attribute.name || field].data;
-      return () => fetch(data);
+      return () => fetch(data).catch((err) => console.error(err));
```
Comment (berfarah) [inline src/model.js]: I need to delete this - I was just using the console log for debugging.

File: src/store.js [hunk 5]
```diff
@@ -1,11 +1,16 @@
+import { createSelector } from 'reselect';
 import './polyfill';
-import { remoteRead } from './api';
+import { apiRead } from './api';
+import { apiIndexAsync, queryEndpoint } from './indexing';
 
 /**
  * @module
  * @private
+ *
```
Comment (berfarah) [inline src/store.js]: It may be worth moving this type definition into model

File: src/model.js [hunk 6]
```diff
@@ -144,22 +144,41 @@ export default function define(type, model = class {}) {
...
+     * @return {RecordPromise<Array<Model>>}
+     */
+    static findAll = (options) => Store.findAll({ ...options, model: this });
+
+    /**
+     * Fetches a query from the server.
+     * @see {@link module:active-redux/store.query}
+     * @return {RecordPromise<Array<Model>>}
+     */
+    static query = (query, options) => Store.query(query, { ...options, model: this });
```
Comment (berfarah) [inline src/model.js]: This totally supports pagination \o/ almost as a side effect

File: src/store.js [hunk 7]
```diff
@@ -16,40 +21,14 @@ export const missingStore = () => (
...
-        return queryValue.includes(entityValue);
-      }
-      if (typeof queryValue === 'object') {
-        return deepPartialEqual(queryValue, entityValue);
-      }
-      return queryValue === entityValue;
-    }, true)
-  ))
-);
+const createRecordPromise = ({ promise, selector, store }) => {
```
Comment (berfarah) [inline src/store.js]: Same thing here about subclassing primitives - I could subclass `Promise` and wrap it, but I think my intentions are fairly clear.

File: src/store.js [hunk 8]
```diff
@@ -16,40 +21,14 @@ export const missingStore = () => (
   `)
 );
 
-// Used for querying the store #where / #find
-const deepPartialEqual = (obj1, obj2) => (
-  Object.entries(obj1).reduce((equal, [key, value]) => {
-    if (equal === false) return equal;
-    if (obj2 === undefined) return false;
+const toModel = (Model) => (data = null) => (data ? new Model(data) : data);
```
Comment (berfarah) [inline src/store.js]: Yay currying!

File: src/store.js [hunk 9]
```diff
@@ -69,39 +48,58 @@ class Store {
...
+      (state) => state.api.indices[endpoint],
+      (state) => state.api.resources[model.type],
+      (index = []) => {
+        const results = index.map((item) => this.peek(item.id, { model }));
+        results.isFetching = index.isFetching;
+        return results;
+      }
+    );
+
+    return createRecordPromise({ promise, selector, store: this.store });
```
Comment (berfarah) [inline src/store.js]: This abstraction feels fairly clean now, I think. It felt pretty forced earlier, but now I feel like things live in the right places. But would love feedback on this.

**Comment by gotno** [comment]
oh, and in your example app setup instructions, there's at least one step missing. you have to `yarn build` in the root to get the `dist` folder in the first place. after symlinking, running `yarn start` gave me

```Failed to compile.

./src/index.js
Module not found: Can't resolve 'active-redux' in '/Users/ont/Dev/active-redux/example/src'```

**Comment by berfarah** [comment]
@gotno Thanks for that - I'll update the example app setup instructions.

To answer your questions:

Relationships are `peek` by default, `fetchX` is a find. I felt it wasn't necessary to implement select, since the usecase (to me) isn't there. Accounts should just have access to existing relationships. Here's an example of how you could implement a dynamic fetch (as needed).

```jsx
import React from 'react';
import { connect } from 'active-redux';
import { Account } from '@models';

class Row extends React.Component {
  constructor() {
    super();
    this.state = { isFetching: false };
  }

  conditionallyFetchOwner() {
    if (!this.state.isFetching && !this.props.account.owner) {
      this.setState({ isFetching: true }, () => this.props.account.fetchOwner());
    }
  }

  render() {
    this.conditionallyFetchOwner();
    return (
      <li>{account.uuid} {account.owner && account.owner.name}</li>
    );
  }
};

const Row = ({ account = {} }) => {
  if (!account.owner) account.fetchOwner()
};

const Table = ({ accounts = [] }) => (
  <ul>
    {accounts.map((account) => <Row key={account.uuid} account={account} />)}
  </ul>
);

const mapStateToProps => () => {
  const accountQuery = Account.selectQuery;
  return (state, props) => {
    accounts: accountQuery({ page: props.page })(state),
  };
};

const mapDispatchToProps = () => ({});

export default connect(mapStateToProps, mapDispatchToProps)(Table);
```

Thanks for your feedback with the observations!

For the index/show endpoints - you're right. There isn't a great way to get around this (at least, the way I have it currently written).

For the `qs` query string - that's troubling. But also totally fixable. I can make it so that at the very least, the top-level keys are sorted before being run through `qs` in a subsequent PR.

For the instructions, I'll add that to the current PR via the README.

**Pull request merged**